src/generate/rust.rs

```diff
@@ -382,10 +487,71 @@ impl<Pat: MatchesEmpty + RustInputPat> GrammarGenerateMethods<Pat> for grammer::
         .parse::<Src>()
         .unwrap();
 
+        let mut records = BTreeSet::new();
+
         for (&name, &rule) in rules.named {
-            out += declare_rule(name, rule, cx, &mut rules) + impl_parse_with(cx, name);
+            let rust_adt = rule.rust_adt(cx, &mut records);
+            out += declare_rule(name, rule, &rust_adt, cx, &mut rules) + impl_parse_with(cx, name);
         }
 
+        let records = records.into_iter().map(|fields| {
+            let ident = Src::ident(&(fields.join("__") + "__"));
+            // FIXME(eddyb) figure out a more efficient way to reuse
+            // iterators with `quote!(...)` than `.collect::<Vec<_>>()`.
```

Comment:
What version of `quote` are we using? IIRC `quote:1` behaves better with iterators now. At the very least we can `let f_ident = fields.iter().map(Src::ident); let f_ident_again = f_ident.clone();`.

Reply:
`proc-quote` and I'd rather not make things *that* ugly.